azure/services/virtualmachines/spec.go

```diff
@@ -277,18 +291,77 @@ func (s *VMSpec) generateOSProfile() (*compute.OSProfile, error) {
 	}
 
-	if !s.SKU.HasCapability(resourceskus.EncryptionAtHost) {
-		return nil, azure.WithTerminalError(errors.Errorf("encryption at host is not supported for VM type %s", s.Size))
+	securityProfile := &compute.SecurityProfile{}
+
+	if storageProfile.OsDisk.ManagedDisk != nil &&
+		storageProfile.OsDisk.ManagedDisk.SecurityProfile != nil &&
+		storageProfile.OsDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType != "" {
+		if s.SecurityProfile.EncryptionAtHost != nil && *s.SecurityProfile.EncryptionAtHost {
+			return nil, azure.WithTerminalError(errors.New("encryption at host is not supported when SecurityEncrytionType is set"))
+		}
+
+		securityProfile.UefiSettings = &compute.UefiSettings{}
+
+		if s.SecurityProfile.SecureBoot != "" && s.SecurityProfile.SecureBoot == infrav1.SecureBootPolicyEnabled {
+			securityProfile.UefiSettings.SecureBootEnabled = pointer.Bool(true)
+		} else {
+			securityProfile.UefiSettings.SecureBootEnabled = pointer.Bool(false)
+		}
+
+		if storageProfile.OsDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType == compute.SecurityEncryptionTypesDiskWithVMGuestState && !*securityProfile.UefiSettings.SecureBootEnabled {
+			return nil, azure.WithTerminalError(errors.Errorf("secure boot should be enabled when SecurityEncrytionType is set to %s", compute.SecurityEncryptionTypesDiskWithVMGuestState))
+		}
+
+		if s.SecurityProfile.VirtualizedTrustedPlatformModule != "" && s.SecurityProfile.VirtualizedTrustedPlatformModule == infrav1.VirtualizedTrustedPlatformModulePolicyDisabled {
+			return nil, azure.WithTerminalError(errors.New("vTPM should be enabled when SecurityEncrytionType is set"))
+		}
+
+		securityProfile.UefiSettings.VTpmEnabled = pointer.Bool(true)
+
+		securityProfile.SecurityType = compute.SecurityTypesConfidentialVM
+
+		return securityProfile, nil
+	}
+
+	if s.SecurityProfile.EncryptionAtHost != nil {
+		if !s.SKU.HasCapability(resourceskus.EncryptionAtHost) && *s.SecurityProfile.EncryptionAtHost {
+			return nil, azure.WithTerminalError(errors.Errorf("encryption at host is not supported for VM type %s", s.Size))
+		}
+
+		securityProfile.EncryptionAtHost = s.SecurityProfile.EncryptionAtHost
 	}
 
-	return &compute.SecurityProfile{
-		EncryptionAtHost: s.SecurityProfile.EncryptionAtHost,
-	}, nil
+	securityProfile.UefiSettings = &compute.UefiSettings{}
+
+	if s.SecurityProfile.SecureBoot != "" {
+		if s.SKU.HasCapability(resourceskus.TrustedLaunchDisabled) && s.SecurityProfile.SecureBoot == infrav1.SecureBootPolicyEnabled {
+			return nil, azure.WithTerminalError(errors.Errorf("secure boot is not supported for VM type %s", s.Size))
+		}
+
+		if s.SecurityProfile.SecureBoot == infrav1.SecureBootPolicyEnabled {
+			securityProfile.UefiSettings.SecureBootEnabled = pointer.Bool(true)
+			securityProfile.SecurityType = compute.SecurityTypesTrustedLaunch
+		}
+	}
+
+	if s.SecurityProfile.VirtualizedTrustedPlatformModule != "" {
+		if s.SKU.HasCapability(resourceskus.TrustedLaunchDisabled) &&
```

Comment:
nit -- calling this twice, and it's a long conditional. create a var? (other call site is the capability check earlier)

Reply:
or combine them?